inline_verifier.go

```diff
@@ -125,20 +212,181 @@ func (v *InlineVerifier) CheckFingerprintInline(tx *sql.Tx, targetDb, targetTabl
 				return nil, fmt.Errorf("cannot convert column %v to []byte", col.Name)
 			}
 
-			sourceDecompressedData[pk][col.Name], err = v.DecompressData(table, col.Name, compressedData)
+			sourceDecompressedData[pk][col.Name], err = v.decompressData(table, col.Name, compressedData)
+		}
+	}
+
+	return v.compareHashesAndData(sourceFingerprints, targetFingerprints, sourceDecompressedData, targetDecompressedData), nil
+}
+
+func (v *InlineVerifier) PeriodicallyVerifyBinlogEvents() {
+	v.logger.Info("starting periodic reverifier")
+
+	for !v.periodicVerifyStopped {
+		time.Sleep(v.VerifyBinlogEventsInterval)
+		_, mismatches, err := v.verifyAllEventsInStore()
+		if err != nil {
+			v.ErrorHandler.Fatal("inline_verifier", err)
+		}
+
+		v.readdMismatchedPKsToBeVerifiedAgain(mismatches)
+
+		v.logger.WithFields(logrus.Fields{
+			"remainingRowCount": v.reverifyStore.currentRowCount,
+		}).Debug("reverified")
+	}
+
+	v.logger.Info("shutdown periodic reverifier")
+}
+
+func (v *InlineVerifier) StopPeriodicallyVerifyBinlogEvents() {
+	v.periodicVerifyStopped = true
+	v.logger.Info("request shutdown of periodic reverifier")
+}
+
+func (v *InlineVerifier) VerifyBeforeCutover() error {
+	var timeToVerify time.Duration
+	// Iterate until the reverify queue is small enough
+	for i := 0; i < v.FinalReverifyMaxIterations; i++ {
+		before := v.reverifyStore.currentRowCount
+		start := time.Now()
+
+		_, mismatches, err := v.verifyAllEventsInStore()
+		if err != nil {
+			return err
+		}
+
+		v.readdMismatchedPKsToBeVerifiedAgain(mismatches)
+		after := v.reverifyStore.currentRowCount
+		timeToVerify = time.Now().Sub(start)
+
+		v.logger.WithFields(logrus.Fields{
+			"store_size_before": before,
+			"store_size_after":  after,
+			"iteration":         i,
+		}).Infof("completed re-verification iteration %d", i)
+
+		if after <= 1000 || after >= before {
```

Comment:
This is how the IterativeVerifier functioned previously. If there's less than 1000 rows in the reverifystore, then it will stop the reverification before cutover and allow cutover to proceed.